csrc/dispatch_utils.h

```diff
@@ -14,15 +14,18 @@
 #define VLLM_DISPATCH_FLOATING_TYPES(TYPE, NAME, ...) \
   AT_DISPATCH_SWITCH(TYPE, NAME, VLLM_DISPATCH_CASE_FLOATING_TYPES(__VA_ARGS__))
 
-// TODO(luka/varun): use FP8_TYPE macro after refactoring
-#ifndef USE_ROCM
-  #define VLLM_DISPATCH_CASE_QUANT_TYPES(...)                    \
-    AT_DISPATCH_CASE(at::ScalarType::Float8_e4m3fn, __VA_ARGS__) \
-    AT_DISPATCH_CASE(at::ScalarType::Char, __VA_ARGS__)
-#else
+// ROCm devices might use either fn or fnuz, so set up dispatch table for both.
+// A host-based check at runtime will create a preferred FP8 type for ROCm
+// such that the correct kernel is dispatched.
+#ifdef USE_ROCM
   #define VLLM_DISPATCH_CASE_QUANT_TYPES(...)                      \
+    AT_DISPATCH_CASE(at::ScalarType::Float8_e4m3fn, __VA_ARGS__)   \
     AT_DISPATCH_CASE(at::ScalarType::Float8_e4m3fnuz, __VA_ARGS__) \
     AT_DISPATCH_CASE(at::ScalarType::Char, __VA_ARGS__)
+#else
```

Comment:
We could add a `#define VLLM_DISPATCH_CASE_FP8_TYPES` here as well, with only one case on CUDA and 2 cases on ROCm

Reply:
I could, but where/how do you see that being used?  

We already have `VLLM_DISPATCH_CASE_QUANT_TYPES(...)` where I added the two fp8 cases on rocm while cuda still has just the one.

Then in [csrc/layernorm_quant_kernels.cu](https://github.com/vllm-project/vllm/blob/04222984f870f3e4ce92076a322e55b18427f40a/csrc/layernorm_quant_kernels.cu#L179-L185), the dispatch is based on the `input.scalar_type()` but we need to also conditionalize the dispatch based on the `out.scalar_type()`.  We're not dispatching based on the fp8 type.  

Did you mean something else?

Reply:
Yes, I imagined nesting the dispatching macros. I've seen it before but we can definitely keep it this way if nesting doesn't work.

Reply:
Nesting dispatch macros was the way to go.